tests/e2e/lib/k8s_common_helpers.go

```diff
@@ -0,0 +1,279 @@
+	}
+
+	if kubeConfig == nil {
+		return "", "", fmt.Errorf("No valid rest.Config provided")
+	}
+
+	if params.PodName == "" {
+		return "", "", fmt.Errorf("No proxy pod specified for the command: %s", command)
+	}
+
+	if params.Namespace == "" {
+		return "", "", fmt.Errorf("No proxy pod namespace specified for the command: %s", command)
+	}
+
+	if params.ContainerName != "" {
+		containerName = params.ContainerName
+	} else {
+		containerName = "curl-tool"
+	}
+
+	log.Printf("Provided command: %s\n", command)
+	log.Printf("Command will run in the pod: %s, container: %s in the namespace: %s\n", params.PodName, containerName, params.Namespace)
+
+	option := &corev1.PodExecOptions{
+		Command:   strings.Split(command, " "),
+		Stdin:     false,
+		Stdout:    true,
+		Stderr:    true,
+		TTY:       true,
+		Container: containerName,
+	}
+
+	postRequest := kubeClient.CoreV1().RESTClient().Post().
+		Resource("pods").
+		Name(params.PodName).
+		Namespace(params.Namespace).
+		SubResource("exec").
+		VersionedParams(option, scheme.ParameterCodec)
+
+	log.Printf("Full command URL: %s\n", postRequest.URL())
+
+	executor, err := remotecommand.NewSPDYExecutor(kubeConfig, "POST", postRequest.URL())
+	if err != nil {
+		return "", "", err
+	}
+
+	stdOutput := &bytes.Buffer{}
+	stdErrOutput := &bytes.Buffer{}
+
+	err = executor.Stream(remotecommand.StreamOptions{Stdout: stdOutput, Stderr: stdErrOutput})
+
+	if stdOutput.Len() > 0 {
+		log.Printf("stdOutput: %s\n", stdOutput.String())
+	}
+
+	if stdErrOutput.Len() > 0 {
+		log.Printf("stdErrOutput: %s\n", stdErrOutput.String())
+	}
+
+	if err != nil {
+		log.Printf("Error while streaming command output: %v\n", err)
+		return stdOutput.String(), stdErrOutput.String(), fmt.Errorf("Error while streaming command output: %v", err)
+	}
+
+	return stdOutput.String(), stdErrOutput.String(), nil
+}
+
+// GetFirstPodByLabel retrieves a first found pod in the specified namespace based on the given label selector.
+// It uses the provided Kubernetes client to interact with the Kubernetes API.
+//
+// Parameters:
+//   - clientset: A pointer to the Kubernetes client (*kubernetes.Clientset).
+//   - namespace: The namespace in which to search for the pod.
+//   - labelSelector: The label selector to filter pods.
+//
+// Returns:
+//   - (*corev1.Pod, error): A pointer to the first pod matching the label selector, or an error if any.
+//
+// Example:
+//
+//	pod, err := GetFirstPodByLabel(clientset, "mongo-persistent", "curl-tool=true")
+//	if err != nil {
+//	    log.Println(err)
+//	}
+//	log.Printf("Pod found: %s\n", pod.Name)
+func GetFirstPodByLabel(clientset *kubernetes.Clientset, namespace string, labelSelector string) (*corev1.Pod, error) {
+
+	podOptions := metav1.ListOptions{
+		LabelSelector: labelSelector,
+	}
+
+	podList, err := clientset.CoreV1().Pods(namespace).List(context.Background(), podOptions)
+
+	if err != nil {
+		return nil, err
+	}
+
+	if len(podList.Items) == 0 {
```

Comment:
should not the check be for length greater then 1?

Reply:
Why?
See the error for that if (no pods, so 0):
```
		return nil, fmt.Errorf("no pods found with label selector: %s", labelSelector)
```

Reply:
If you get more than 1 result, command may end up executing in different pods during runs, which can create unreproducible behaviors?

Reply:
No, that is using route from any container within cluster that is labeled, we don't care which one. The function name is pretty precise and I have change the function name to address comment from @kaovilai, because at first was not clear that's it's first matching pod.


```
// GetFirstPodByLabel retrieves a first found pod in the specified namespace based on the given label selector.
// It uses the provided Kubernetes client to interact with the Kubernetes API.
//

// Returns:
//   - (*corev1.Pod, error): A pointer to the first pod matching the label selector, or an error if any.
func GetFirstPodByLabel
```

Reply:
> it will use route from any container within cluster that is labeled as a proxy

Do you mean it will only get pods with label curltool which is a deployment we control so we would limit replicas to 1 anyway? Or if they are multiple replicas, the application would be mounting the same PVC thus returning the same data for volume persistence tests.

Reply:
@kaovilai yes we do control deployment and we limit to 1 anyway, but imo it's always better in a function act to return only what is needed that's why there is limit of 1. To be fair even if you have multiple deployments in one namespace and for example left overs from other jobs or additionally deployed apps it will still work. Within the same namespace we really don't care who will act as a proxy as it will always find it's way via internal routing to the service within the same namespace.

https://github.com/openshift/oadp-operator/pull/1253/files#diff-454849abc186dbe835e4d394e7142c24d4afcc776d63d556bae8df350db2d723R66

If the user has access to other namespaces the above statement will be valid across multiple namespaces as we try to reach the endpoint using it's service name and internal hostname:

https://github.com/openshift/oadp-operator/pull/1253/files#diff-f56b125140f13a9f4b7a6959d7a2224a1368223762f064fe7c1490729a490f0eR147